### PR TITLE
[WIP] Support of C++ list initializers in Python using tuples

### PR DESCRIPTION
**Pull request by peremato**
The idea here is to allow users to provide a python tuple to implicitly construct and initialize an object that is needed in a function or method argument. Instead of doing the flowing when calling a function/methods that requires `const A&`
```
a = ROOT.A(arg1,arg2) 
ROOT.function(a) 
```
the user can do 
```
ROOT.function((arg1,arg2))
```

### CODE REVIEW

File: bindings/pyroot/src/Converters.cxx [hunk 1]
```diff
@@ -1013,24 +1013,54 @@ Bool_t PyROOT::TValueCppObjectConverter::SetArg(
...
+      }
+   }
+   else if ( PyTuple_Check( pyobject ) ){  // It is a Python tuple (equivalent to a C++ initializer list)
+
+      // instantiate an object proxy of this class
+      if( ! fObjProxy ) {
+      // retrieve python class
+         PyObject* pyclass = CreateScopeProxy( fClass );
+         if ( ! pyclass ) return kFALSE;                  // error has been set in CreateScopeProxy
+         fObjProxy = (ObjectProxy*)((PyTypeObject*)pyclass)->tp_new( (PyTypeObject*)pyclass, pyobject, NULL );
```
Comment (etejedor) [inline bindings/pyroot/src/Converters.cxx]: Do we need to pass the tuple (pyobject) when we construct the proxy here?

File: bindings/pyroot/src/Converters.cxx [hunk 2]
```diff
@@ -1013,24 +1013,54 @@ Bool_t PyROOT::TValueCppObjectConverter::SetArg(
...
+      Py_DECREF( obj );
+
+      para.fValue.fVoidp = fObjProxy->GetObject();
       para.fTypeCode = 'V';
-      return kTRUE;
-   } else if ( ! TClass::GetClass( Cppyy::GetFinalName( fClass ).c_str() )->GetClassInfo() ) {
-   // assume "user knows best" to allow anonymous reference passing
-      para.fValue.fVoidp = pyobj->GetObject();
-      para.fTypeCode = 'V';
+      // TODO we need to remove the temporary object constructed from the tuple
```
Comment (etejedor) [inline bindings/pyroot/src/Converters.cxx]: This is done in lines 1047 - 1049

File: bindings/pyroot/src/Converters.cxx (hunk 1 above)
Comment (peremato) [reply bindings/pyroot/src/Converters.cxx]: No, it not needed. The last two arguments are ignored. We can call it as `tp_new( (PyTypeObject*)pyclass, NULL, NULL )`

File: bindings/pyroot/src/Converters.cxx (hunk 2 above)
Comment (peremato) [reply bindings/pyroot/src/Converters.cxx]: Yes. The comment can be deleted.

Comment (dpiparo) [reply bindings/pyroot/src/Converters.cxx]: Yes sure! I remembered that we needed to have this put aside as we did for strings :)
Thanks.

File: bindings/pyroot/test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ROOT_ADD_TEST(pyroot-list-init 
+              COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/list-initialization.py)
```
[the remark this answers is not in the document]
Comment (peremato) [reply bindings/pyroot/test/CMakeLists.txt]: done

File: bindings/pyroot/test/list-initialization.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        self.assertEqual(ROOT.get_i((10,20.)), 10)
+        self.assertEqual(ROOT.get_f((10,20.)), 20.)
+        self.assertAlmostEqual(ROOT.get_f((10,)), 99.99)
+    
+    def test_invalid_constructor(self):
+        with self.assertRaises(TypeError): ROOT.get_i((1,2,3))
+        with self.assertRaises(TypeError): ROOT.get_i((1,'abc'))
+
+if __name__ == '__main__':
+    unittest.main()
```
[the remark this answers is not in the document]
Comment (peremato) [reply bindings/pyroot/test/list-initialization.py]: done

**Comment by peremato** [comment]
@Axel-Naumann do you refer to std::initializer_list?

**Comment by Axel-Naumann** [comment]
Yes, `std::initializer_list`. It's sufficiently unrelated from list init; I'm not suggesting to add it to this PR.

**Comment by dpiparo** [comment]
@phsft-bot build, please.

**Pull request merged**